### PR TITLE
Returning `undefined` for options as promised in the signature

### PR DESCRIPTION
**Pull request by robertbastian**

### CODE REVIEW

**Comment by Manishearth** [comment]
Should we? `undefined` is kinda weird and annoying, `null` does actually seem better to me, especially since we treat None objects as `null`.

**Comment by sffc** [comment]
[ECMA-262](https://tc39.es/ecma262/#sec-undefined-value) says:

> 4.4.13 undefined value
> primitive value used when a variable has not been assigned a value
>
> 4.4.15 null value
> primitive value that represents the intentional absence of any object value

[MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#undefined_type) says:

> Conceptually, undefined indicates the absence of a value, while null indicates the absence of an object
> ...
> Many methods, such as [Array.prototype.find()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find) and [Map.prototype.get()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/get), return undefined when no element is found.
> ...
> null is used much less often in the core language. The most important place is the end of the [prototype chain](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Inheritance_and_the_prototype_chain)

In the spec, `null` and `undefined` are often treated in the same way, except `null` is used a lot in prototype-related things.

So maybe `null` is like `()` and `undefined` is like `None`.

File: tool/src/js/conversions.rs
Comment (sffc) [inline tool/src/js/conversions.rs:458]: Thought: You could also do something like

```
let rv;
if ({option_ptr} == 0) rv = {};
return rv;
```

(not necessarily advocating for this; what you have is fine and nice that it doesn't need a local variable)

**Review by sffc — APPROVED** [review]

**Review by Manishearth — CHANGES_REQUESTED** [review]
(please don't merge yet, I'm still not sure of this and will write down thoughts tomorrow)

**Review by Manishearth — CHANGES_REQUESTED** [review]
My three concerns are this:

Firstly, in TypeScript, `?` means null except in struct field contexts (where it tacks on an implicit `| undefined`. We're not using `?` here, which I guess is a choice. `?` is definitely more convenient to work with in TS but undefined isn't too bad.

Secondly, we use `null` with object types, and we should continue to do so. We should perhaps have a test that ensures that `-> Option<OpaqueType>` _always_ produces a nullable type rather than a nullable _return_. I'd never want that situation to return `undefined`. I would ideally prefer this to be uniform as well; so people don't have to worry about whether a Diplomat nullable is `null` or `undefined`. Down this path lies the den of the `==` operator.

The third one is the one I haven't yet reconciled; at least in WebIDL nullable types when produced [are treated as "or null` regardless of the kind of type](https://webidl.spec.whatwg.org/#js-nullable-type):

> The result of [converting](https://webidl.spec.whatwg.org/#dfn-convert-idl-to-javascript-value) an IDL [nullable type](https://webidl.spec.whatwg.org/#dfn-nullable-type) value to a JavaScript value is:
>
> If the IDL [nullable type](https://webidl.spec.whatwg.org/#dfn-nullable-type) T? value is null, then the JavaScript value is null.


So there does seem to be a tradition of using `valueType?` to mean `valueType | null`


Personally my experience with JS is that `undefined` really does mean "not defined", _not_ a sentinel value. The docs Shane linked match that, I think this falls under the scope of "primitive value that represents the intentional absence of any object value".

I have only ever seen `undefined` used for cases like "the caller didn't set this" or for leaving _object fields_ unset.

I disagree with this characterization:

> So maybe null is like () (an empty object) and undefined is like None (the absence of a value).

I don't think it matches what the docs say, and `()` is more like a declaration that there will never be an object here. `None` is an intentional signal of an absence of a value.



I don't actually care much for JS1, but we should decide on this for JS2.

**Comment by robertbastian** [comment]
I think this discussion is misplaced. We already have undefined on the signature, this PR just aligns the implementation. I'm happy to change the signature in JS2.

**Review by Manishearth — APPROVED** [review]
I'd rather chanfe JS1 but I'm fine with this then

**Comment by sffc** [comment]
Some opinions on Matrix:

https://matrix.to/#/!wbACpffbfxANskIFZq:matrix.org/$KoVWZpZ_HbZViQstfU7261-zho71g7gZtAw1RrT6kTE?via=matrix.org&via=mozilla.org&via=igalia.com

**Pull request merged**
